src/bootstrap/doc.rs

```diff
@@ -597,7 +600,9 @@ impl Step for Rustc {
         t!(fs::create_dir_all(&out));
         let compiler = builder.compiler(stage, build.build);
         let rustdoc = builder.rustdoc(compiler.host);
+        let mut dylib_path = vec![];
         let compiler = if build.force_use_stage1(compiler, target) {
+            dylib_path.push(builder.rustc_libdir(compiler));
```

Comment:
Hm could you detail a bit why this (and the above one) is needed? Shouldn't the rpath configuration cover this?

Reply:
It builds with rpath disabled because with rpath enabled it ends up with long library path with `..` directories.

Reply:
Hm when you say "it" in "it builds with rpath disabled" what is that referring to?

Reply:
I mean Rust installation script which builds Rust from source.

Reply:
I believe the rpath should be [configured here](https://github.com/rust-lang/rust/blob/a3a7203e2c9ed30a501da86f3fa1f9efe707ac94/src/bootstrap/bin/rustc.rs#L191-L235) and IIRC that's enabled by default, but are you disabling that?

Reply:
Yes, it disabled: https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-lang/rust/rust-1.19.0.ebuild#n107
As I see Ubuntu also build rust same way.

Reply:
Ok thanks for the info! I'm also trying to page all this back in and recall all what's going on here. IIRC we specifically *don't* change the dylib path of the `cargo` command here to get things working in stage0 (in theory) where the libpath is different than it is in later stages.

For example I believe that's what's going on [here](https://github.com/rust-lang/rust/blob/a3a7203e2c9ed30a501da86f3fa1f9efe707ac94/src/bootstrap/bin/rustc.rs#L76-L85) where we select, at build time, which lib path we should be using.

Is the problem that this change here fixing that building documentation fails? If so that's when executing `rustdoc` the dylibs can't be found, right? If so this *should* be fixed by [this line](https://github.com/rust-lang/rust/blob/a3a7203e2c9ed30a501da86f3fa1f9efe707ac94/src/bootstrap/bin/rustdoc.rs#L26) but I'm not sure why that's not sufficient.

Reply:
> For example I believe that's what's going on here where we select, at build time, which lib path we should be using.

I didn't look here. I wonder if `rustc` related with issue because failed command is `rustdoc`.

> Is the problem that this change here fixing that building documentation fails? If so that's when executing rustdoc the dylibs can't be found, right? If so this should be fixed by this line but I'm not sure why that's not sufficient.

It sets path to stage1 while we launch stage2 rustdoc

Reply:
> For example I believe that's what's going on here where we select, at build time, which lib path we should be using.

I'll try to move this code from `rustc` to `rustdoc` and see what happens without setting dylib.

Reply:
No, its just change path to libdir but there need to be stage1 libdir to generate docs and stage2 libdir for rustdoc stage2 executable.

Reply:
I've tried to set `RUSTC_LIBDIR` to stage2 libdir but https://github.com/rust-lang/rust/blob/master/src/bootstrap/doc.rs#L445 use stage1 rustc so it fails.

Reply:
I wonder if it is better to send libdir for stage2 in environment variable `RUSTDOC_LIBDIR` or so instead set it with `add_lib_path`?

Reply:
Sure yeah adding `RUSTDOC_LIBDIR` sounds good! I think it'd be best to reuse what's already in the shim executables instead of adding more paths here, but hopefully isn't too bad to add!

Reply:
Reimplemented this with `RUSTDOC_LIBDIR`.